### PR TITLE
fix(DOGE/DASH): transfering funds doesn't work after `bitcoinjs-lib` upgrade

### PR DESCRIPTION
**Pull request by RealGoodProgrammer**

### CODE REVIEW

**Review by bludnic — CHANGES_REQUESTED** [review]

File: src/lib/bitcoin/dash-api.js [hunk 1]
```diff
@@ -16,55 +16,61 @@ class DashApiError extends Error {
...
-      .then(result => this._mapTransaction(result))
+  getTransaction(txid) {
+    return this._invoke('getrawtransaction', [txid, true]).then((result) =>
+      this._mapTransaction(result)
+    )
   }
 
   /** @override */
-  getTransactions (options) {
+  async getTransactionHex(txid) {
```
Comment (bludnic) [inline src/lib/bitcoin/dash-api.js:45]: Remove async. Its optional.

File: src/lib/bitcoin/doge-api.js [hunk 2]
```diff
@@ -140,6 +140,12 @@ export default class DogeApi extends BtcBaseApi {
     return this._get(`/api/tx/${txid}`).then((tx) => this._mapTransaction(tx))
   }
 
+  /** @override */
+  getTransactionHex(txid) {
+    const { rawtx } = this._get(`/api/rawtx/${txid}`)
```
Comment (bludnic) [inline src/lib/bitcoin/doge-api.js:145]: You probably need to `await` here

File: src/lib/bitcoin/doge-api.js [hunk 3]
```diff
@@ -3,68 +3,184 @@ import qs from 'qs'
...
+    }
+
+    const hex = await this._buildTransaction(address, amount, unspents, fee)
+
+    let txid = bitcoin.crypto.sha256(Buffer.from(hex, 'hex'))
+    txid = bitcoin.crypto.sha256(Buffer.from(txid))
+    txid = txid.toString('hex').match(/.{2}/g).reverse().join('')
+
+    return { hex, txid }
+  }
```
Comment (bludnic) [inline src/lib/bitcoin/doge-api.js:61]: Can be removed. This function is inherited from btc-base-api

**Review by bludnic — CHANGES_REQUESTED** [review]

File: src/lib/bitcoin/doge-api.js [hunk 4]
```diff
@@ -3,68 +3,178 @@ import qs from 'qs'
...
+
+    txb.addOutput({
+      address,
+      value: amount
+    })
+
+    // Estimated fee based on https://github.com/dogecoin/dogecoin/blob/master/doc/fee-recommendation.md
+    const currentFeeRate = await this.getFeePerByte()
+    let estimatedFee = Math.floor(
+      new BigNumber(currentFeeRate * (estimatedTxBytes + OUTPUTS_COMPENSATION))
```
[the remark this answers is not in the document]
Comment (RealGoodProgrammer) [reply src/lib/bitcoin/doge-api.js:103]: I'll leave this unchanged for better readability

File: src/lib/bitcoin/doge-api.js [hunk 5]
```diff
@@ -3,68 +3,184 @@ import qs from 'qs'
...
+    return { hex, txid }
+  }
+
+  /** @override */
+  async _buildTransaction(address, amount, unspents, fee) {
+    amount = new BigNumber(amount).times(this.multiplier).toNumber()
+    amount = Math.floor(amount)
+    const heldFee = Math.floor(new BigNumber(fee).times(this.multiplier).toNumber())
+
+    const txb = new bitcoin.Psbt({
```
[the remark this answers is not in the document]
Comment (RealGoodProgrammer) [reply src/lib/bitcoin/doge-api.js:69]: I

[the remark this answers is not in the document]
Comment (RealGoodProgrammer) [reply src/lib/bitcoin/doge-api.js:69]: That means "builder". But you're right, it's more appropriate to call it psbt

File: src/lib/bitcoin/doge-api.js [hunk 6]
```diff
@@ -3,68 +3,165 @@ import qs from 'qs'
...
+      if (transferAmount >= target) {
+        break
+      }
+      const buffer = Buffer.from(tx.txHex, 'hex')
+      psbt.addInput({
+        hash: tx.txid,
+        index: tx.vout,
+        nonWitnessUtxo: buffer
+      })
+      transferAmount += tx.amount
```
[the remark this answers is not in the document]
Comment (bludnic) [reply src/lib/bitcoin/doge-api.js:70]: For now I think it's enough to use multiply and divide

**Review by bludnic — APPROVED** [review]

**Comment by bludnic** [comment]
[Update bitcoinjs-lib](https://trello.com/c/IO8SGz5y/322-update-bitcoinjs-lib)

**Pull request merged**